### ISSUE TITLE
Correctness bugs and precision loss with Python-native-floating-point-number/single-valued-tensor divisions

### ISSUE THREAD

**Issue by X-Engineer-001** [pytorch/pytorch#128401]
### 🐛 Describe the bug

```
import torch
def test(a,b):
    print()
    print(a/b)
    print(torch.tensor(a)/b)
    print(a/torch.tensor(b))        # Potential bug.
    print(a/float(torch.tensor(b))) # Potential much more precision loss than merely float() casting.
    print(a/float(b))
test(6.10840013583831e-41, 2.7734e-39)  # Bug to inf. (original found case)
test(0.0, 2.8026e-45)                   # Bug to nan. (original found case)
test(1.0e-39, 1.0e-39)                  # Bug to inf and more precision loss.
test(1.0e-38, 1.0e-39)                  # Bug to inf and more precision loss.
test(1.0e-39, 1.0e-38)                  # More precision loss.
test(1.0e-38, 1.0e-38)                  # More precision loss.
test(0.0, 1.0e-39)                      # Bug to nan.
test(0.0, 1.0e-38)
```
```
0.022024951813075323
tensor(0.0220)
tensor(inf)
0.022024955991519653
0.022024951813075323

0.0
tensor(0.)
tensor(nan)
0.0
0.0

1.0
tensor(1.)
tensor(inf)
0.9999997846947131
1.0

10.0
tensor(10.0000)
tensor(inf)
9.99999784694713
10.0

0.09999999999999999
tensor(0.1000)
tensor(0.1000)
0.10000000649543637
0.09999999999999999

1.0
tensor(1.)
tensor(1.)
1.0000000649543639
1.0

0.0
tensor(0.)
tensor(nan)
0.0
0.0

0.0
tensor(0.)
tensor(0.)
0.0
0.0
```
Accidentally found this. It can be a severe bug or not depending on your opinion.
In my opinion, it's a potential very severe problem, especially the correctness bug.
People don't 100% of the time casting everythings manually in the codes/scripts.
In my case, I first had to notice the nan and inf in my script output, then do some debug to find this problem.
What's worse, if my research had been supposed to contain nan and inf in the output, I would not find this issue and take it as correct results.
It will be disastrous if someone is conducting much more complicated researches and encounter this. Even if they find the issue, they'll have nightmare debugging (given that they don't know this issue from Pytorch tensors).

### Versions

Collecting environment information...
PyTorch version: 2.3.1+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Home
GCC version: (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22631-SP0
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060 Laptop GPU
Nvidia driver version: 546.80
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=2500
DeviceID=CPU0
Family=207
L2CacheSize=11776
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=2500
Name=12th Gen Intel(R) Core(TM) i9-12900H
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pytorch-lightning==1.1.0
[pip3] torch==2.3.1
[pip3] torchaudio==2.2.1+cu118
[pip3] torchmetrics==1.4.0.post0
[pip3] torchvision==0.18.1
[conda] Could not collect

**Comment by colesbury** [comment]
This is not a bug - it's how floating point operations work. `float` in Python is a 64-bits, while the default floating point format for PyTorch tensors is 32-bits.

See:
* https://pytorch.org/docs/stable/notes/numerical_accuracy.html
* https://en.wikipedia.org/wiki/Floating-point_arithmetic
